### DISCUSSION TITLE
Opening links in books/documents in the native web browser

### DISCUSSION THREAD

**Discussion by peterrus** [koreader/koreader#9057] [Ideas]
Using KOReader on a Pocketbook HD3:  When I tap on a link in a `.epub` I get the following dialog:

I wonder if it would be possible to open those links in the device's web browser. Is this a limitation of the Pocketbook implementation currently?

![scr0002](https://user-images.githubusercontent.com/1117858/166215121-23a3e3e1-1d45-406a-9db7-f900c121d654.png)

**Comment by pazos** [comment]
> Is this a limitation of the Pocketbook implementation currently?

Yes.

Platforms that support opening links in a browser are currently Android, Linux and MacOS.

No idea if Pocketbook has an API for that. In that case @rjd22 might be interested in this. It is a matter of implementing `openLink` with specific pocketbook stuff, like we're doing on [android](https://github.com/koreader/koreader/blob/master/frontend/device/android/device.lua#L91-L95) and [desktop](https://github.com/koreader/koreader/blob/master/frontend/device/sdl/device.lua#L72-L77)

**Comment by rjd22** [comment]
@pazos while there might (need to check) be an API for it. Switching back to koreader will not be trivial, making this feature kind of unpleasant to work with most likely.

**Comment by roygbyte** [comment]
Would it not be possible to run a HTTP get request on the link and show the results in a pop-up scroll view, like is done for Wikipedia? From my work on News downloader it seems to me that all that plugin does anyway is fetch a page and put the HTML into an epub. So in this case, we just download the page a put the HTML into the appropriate KOReader view.

**Reply by poire-z** [reply]
Wikipedia gets plain text properly cleaned up from the API.
NewsDownloader should get plain text or simple HTML from the rss/atom feed, hopefully tidy'ed up by the author/web site.
But for web pages in the wild, you'll get not clean/simple HTML, and you may end up showing some CSS/Javascript or complex page layout crap (ie. headers, menu, tables...) before any real valuable content (and so, the user woud have to scroll dozens of page of crap). And it's possible that go get to display anything, the page depends on Javascript fetching content, etc...
So, may work, but most often disgusting result.
So, you'd probably need to pass the content thru a cleaner (like "readability" bookmarklets, or Firefox reader view) which most often depends on having Javascript (because  [TO͇̹̺ͅƝ̴ȳ̳ TH̘Ë͖́̉ ͠P̯͍̭O̚​N̐Y̡ H̸̡̪̯ͨ͊̽̅̾̎Ȩ̬̩̾͛ͪ̈́̀́͘ ̶̧̨̱̹̭̯ͧ̾ͬC̷̙̲̝͖ͭ̏ͥͮ͟Oͮ͏̮̪̝͍M̲̖͊̒ͪͩͬ̚̚͜Ȇ̴̟̟͙̞ͩ͌͝S̨̥̫͎̭ͯ̿̔̀ͅ](https://stackoverflow.com/a/1732454) ).
Last time I checked, I didn't find any Lua HTML cleaner.

**Reply by roygbyte** [reply]
Gosh, by asking alone I fear I've unlocked some demonic entity from the deep web... How could I be so careless 🥲 (that is a beautiful StackOverflow post—definitely going in my "code poetry" bookmarks).

I think Newsdownloader literally just grabs the web page content, though, and it's not always cleaned up. Or maybe it looks for a `main` element, etc. At the very least I'll play around with this possibility ( "unholy radiance destroying enlightenment") because it is intriguing. I'll use the XML parser, though 😛

**Reply by Frenzie** [reply]
You can very easily invoke a ScrollHTMLWidget, and in newer MuPDF than we're current using it could even deal with broken HTML a lot better, but as stated it often won't be particularly useful.

As an aside, there is https://github.com/msva/lua-htmlparser

**Reply by poire-z** [reply]
Oh, I thought NewsDownloader was just dealing with RSS/Atom feeds, and got/show content only from these XML (so, possibly only the introduction/summary of each article). So, you mean it does not do that, but go grab each target real full page for each url referenced in the rss xml ?
If yes, well, ok, you're already meeting the world wild web.

**Reply by rjd22** [reply]
I asked PocketBook if they can supply me with a newer version of the Inkview library. The one we're using was last changed in 2015 and I suspect that we're missing a lot of new methods.

There might be something like OpenBrowser since there are also things like OpenNotepad and OpenPlayer in the current inkview library.

**Reply by Frenzie** [reply]
@poire-z 
> Oh, I thought NewsDownloader was just dealing with RSS/Atom feeds, and got/show content only from these XML (so, possibly only the introduction/summary of each article). So, you mean it does not do that, but go grab each target real full page for each url referenced in the rss xml ?

It can do either one depending on configuration. To my annoyance I think it defaults to grabbing the whole page but then I'm not an active user.

**Reply by roygbyte** [reply]
@poire-z that's what it does, yes. Some feeds do make use of the description element to put the full content, but others require actually visiting the page. So both need to be supported anyways, meaning we are effectively fetching webpages, yes!

**Reply by roygbyte** [reply]
@Frenzie yes I suppose it should use the description content if it's available. But then it would have to make the full request anyways to check that the description contains all the content, wouldn't it?

**Reply by Frenzie** [reply]
@roygbyte Whether it does per-entry page downloading should be based exclusively on if the user tells that feed to do per-entry page downloading. Afaik it works perfectly fine, my only gripe being that the default is incorrect and that you have to disable it.

**Reply by roygbyte** [reply]
@Frenzie right, so if instructed to download the whole feed it ignores the description/summary element and gets the whole thing.

**Reply by Frenzie** [reply]
It could be put in the EPUB summary field or a kind of ToC with the title + summary or something, but ignore for the purposes of the actual content, yes.

**Reply by roygbyte** [reply]
What is the use case of someone not wanting to download the whole feed? I really can't understand it, tbh.

**Reply by Frenzie** [reply]
I can think of some, but that's neither here nor there. :-)

Here's a typical feed: <https://github.com/koreader/koreader/releases.atom>. The majority of feeds looks just like it, I'd say 9 in 10 or more. That's based on the dozens of feeds I subscribe to (not in NewsDownloader), which is of course not a random sample, so it might not be representative. However, it's my interests that make it non-random, not whether they're "full" feeds or not. There could be a correlation between the kind of blog/subject matter I like and "full" feeds.

But let's assume for the sake of argument that it's only a minority, say 3 in 10. Even then it's a much better default to quickly know that it doesn't work out for you, flip the switch to download the linked page instead, and done. The reverse experience is always worse. You'd have to turn it off to find out whether or not the feed is notification only or whether you could save a half a minute not needlessly downloading a million individual pages.

The tailor-made feed content is always better than some random webpage that probably doesn't even load anything at all without 50 megabytes of React.

**Reply by roygbyte** [reply]
@Frenzie yes so I was thinking about this problem before. Like: wanting to see what the feed looks like on device before committing, so to speak, to a subscription. It seems to me that the answer could be in the default behavior of the plugin, like you indicated, where it defaults to a summary. I think another answer would be to provide the user with a "peek" at the feed items. Incidentally, such a functionality would basically accomplish the feature requested by this discussion, which is to get and show webpages in a html scroll view or similar. So we've come full circle, in a way 😛

Also tagging #9225 to remember this thread.

**Reply by Frenzie** [reply]
> It seems to me that the answer could be in the default behavior of the plugin, like you indicated, where it defaults to a summary.

It doesn't default to a summary! It defaults to the content provided by the feed which is *almost never* just a summary. But it would be the superior experience even if it usually were because as long as that scenario isn't exceedingly rare, the experience in downloading the feed with full content is just so much better, because it takes only a fraction of the time and the end result is just that much better.

**Reply by roygbyte** [reply]
What I mean is that at least one feed syndication format (Atom, I believe) only provides an excerpt of the feed entry content in the form of a summary. See [Our World in Data](https://ourworldindata.org/atom.xml) for an example. With respect to RSS, yes, many feeds will put the entire entry's content in the entry's `description` attribute. My own feed does not (because my blog is in org-mode and I don't know how to do that). In these instances, it's necessary to visit the entry's id, url, or permalink (depending on the syndication format) to get the entry content.

Is that what you're saying? I find the nomenclature of feeds to be confusing, especially since they use different terms for the different pieces of content. Maybe you and I are saying the same thing but with different words :') `summary == description == content` ?

So, the default configuration for #9225 Newsdownloader should be to fetch the entry's content by visiting the actual page... is that what you're saying?

**Reply by Frenzie** [reply]
> What I mean is that at least one feed syndication format (Atom, I believe) only provides an excerpt of the feed entry content in the form of a summary.

It most certainly does not, and such a thing does not exist.

> So, the default configuration for https://github.com/koreader/koreader/issues/9225 Newsdownloader should be to fetch the entry's content by visiting the actual page... is that what you're saying?

No, no, no, no, abso-freaking-lutely not! :rofl: I don't know how else to express just how much it should never ever do that because the user experience is atrocious.

**Reply by Frenzie** [reply]
To emulate the experience on a slow device with not so great wifi, try something like this:
```
sudo tc qdisc add dev enp7s0 root tbf rate 50kbit latency 50ms burst 1540
```

Remove it again later when you want your internet to stop being awful.
```
sudo tc qdisc del dev enp7s0 root tbf rate 50kbit latency 50ms burst 1540
```

**Reply by Frenzie** [reply]
> Is that what you're saying? I find the nomenclature of feeds to be confusing, especially since they use different terms for the different pieces of content. Maybe you and I are saying the same thing but with different words :') summary == description == content ?

Atom differentiates between summary and content. Unfortunately Atom does require something in there to be valid, making notification-only feeds invalid without something silly like a `.` in there.

RSS doesn't differentiate, only offering a description (which sounds like summary, but is used for both).

But my basic point is that figuring out you need to switch to visiting the actual page takes seconds. Figuring out that you'll need to change to just using the feed as is because it's the vastly superior option is something you probably wouldn't even think to try, not realizing it takes 30 times longer per feed for worse but probably usable results.

**Reply by roygbyte** [reply]
Re: opening links

What about: add a "download as EPUB" button to the menu that pops up (alongside "Add to Wallabag") and then just download it like we do for Newsdownloader?

**Reply by Frenzie** [reply]
In between QR code and Wallabag I'm covered personally, but other than making the UI too busy I don't see much in favor or against.

**Reply by roygbyte** [reply]
It would be a good stop gap for lack of a web browser. Obviously not a perfect solution, and could create lots of litter along the way.... But wallabag would do that too. I suppose a person should be able to choose what kind of way a link is executed (add to wallabag, output to epub, etc)...

**Comment by rjd22** [comment]
To also give any update on the Inkview support for opening the browser.

The awesome people of PocketBook made an update to their SDK repository for us to look at:
https://github.com/pocketbook/SDK_6.3.0/tree/6.5

Their updated Inkview library can be found here:
https://github.com/pocketbook/SDK_6.3.0/blob/6.5/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h

Sadly there is no support for opening a link in the PocketBook browser. I possibly could ask them if they could add it for us but IMO it would then take years before we could safely use it because of old devices.

So sadly no good news there.

**Comment by dmalinovsky** [comment]
Looks like it's fixed in https://github.com/koreader/koreader/pull/11787.
